.github/workflows/ci.yml

```diff
@@ -16,23 +16,35 @@ env:
 jobs:
   test:
     name: Test ${{ matrix.crate }}
-    runs-on: ubuntu-latest
+    runs-on: ${{ fromJSON(
+      github.repository == 'libp2p/rust-libp2p' && (
+        (contains(fromJSON('["libp2p-webrtc", "libp2p"]'), matrix.crate) && '["self-hosted", "linux", "x64", "2xlarge"]') ||
+        (contains(fromJSON('["libp2p-quic", "libp2p-perf"]'), matrix.crate) && '["self-hosted", "linux", "x64", "xlarge"]') ||
+        '["self-hosted", "linux", "x64", "large"]'
```

Comment:
I think it was ~+8m for `libp2p-webrtc` and `libp2p` and ~+4m for `libp2p-quic` and `libp2p-perf` which would take the entire workflow to the region of ~18m (all running on `large`).

In terms of vCPUs, `large` gets `2`, `xlarge` is `4` and `2xlarge` is `8`. We start ~40 jobs per workflow here. If they were all on `2xlarge`, we'd use `320` vCPUs per run. We currently get ~1200 vCPUs limit in that AWS account. So it'd take 4 runs to exhaust that. Another thing to consider is how many machines of each kind are generally available at a time (we're operating in `us-east-1`) - I haven't checked it but I guess we could start having to wait for instance type availability. 

In terms of time/money, it'd be worth checking things more closely. It might be the case that it's worth getting more powerful machines for some other workflows to finish them more quickly. Less runtime = lower cost. This would take a bigger analysis though.

In the future, I'd like to get better insights into resource usage per workflow job. This would help us choose the instance types better. We're not there yet.

Reply:
Cool, thanks for the nice breakdown!

It sounds to me like picking the right machine for each job is definitely worth it here.

Once we cut the next round of breaking releases, we can also [bump the MSRV](https://github.com/libp2p/rust-libp2p/issues/3787) to be the same for all crates which should reduce the MSRV-cache size a fair bit and cut the download time there too, making these jobs even faster.

Reply:
Not a blocker, just thinking out loud:

We already have a CI step to dynamically compute the list of crates. I wonder if it is possible to merge the spec of what runner to run a crate on with that such that we only need to refer to here instead of being conditional on the name.